drivers/src/pmu.rs

```diff
@@ -0,0 +1,134 @@
+        for i in 0..num_counters.min(MAX_HARDWARE_COUNTERS as u64) {
+            if matches!(pmu_info.init_counter(i)?, ControlFlow::Break(())) {
+                break;
+            }
+        }
+        if !pmu_info.hw_counters_info.is_empty() {
+            PMU_INFO.call_once(|| pmu_info);
+            Ok(())
+        } else {
+            Err(SbiError::NotSupported)
+        }
+    }
+
+    /// Returns a reference to the global PmuInfo structure.
+    pub fn get() -> SbiResult<&'static PmuInfo> {
+        PMU_INFO.get().ok_or(SbiError::NotSupported)
+    }
+
+    /// Returns the counter index for the specified CSR
+    pub fn csr_to_counter_index(&self, csr: u64) -> SbiResult<u64> {
+        let counter_index = self
+            .hw_counters_info
+            .iter()
+            .position(|info| info.filter(|i| i.get_csr() == csr).is_some())
+            .ok_or(SbiError::InvalidParam)? as u64;
+        Ok(counter_index)
+    }
+
+    /// Returns the number of hardware counters supported by the platform
+    pub fn get_num_counters(&self) -> u64 {
+        self.hw_counters_info.len() as u64
+    }
+
+    /// Returns cached information for the hardware counter specified by counter_index.
+    pub fn get_counter_info(&self, counter_index: u64) -> SbiResult<PmuCounterInfo> {
+        let info = self
+            .hw_counters_info
+            .get(counter_index as usize)
+            .ok_or(SbiError::InvalidParam)?;
+        if let Some(info) = info {
+            Ok(*info)
+        } else {
+            Err(SbiError::InvalidParam)
+        }
+    }
+
+    /// Returns the CSR for the specified counter index
+    pub fn counter_index_to_csr(&self, counter_index: u64) -> SbiResult<u64> {
+        Ok(self.get_counter_info(counter_index)?.get_csr())
+    }
+
+    /// Validates the counter_index and counter_mask bit-mask for counter start and stop operations.
+    /// The counter_mask is relative to counter_index. The return value is a sanitized version
+    /// of the original counter_mask, with set bits representing implemented platform counters.
+    pub fn validate_counter_and_mask(
+        &self,
+        counter_index: u64,
+        counter_mask: u64,
+    ) -> SbiResult<u64> {
+        let counter_index = counter_index as usize;
+        // Validate entire counter mask to ensure that no spurious bits corresponding to
+        // non-hardware counter indexes were set. Since hardware counters can be sparse, it's OK to set bits
+        // for unimplemented hardware counters, as long as they don't exceed the last known valid index. The
+        // platform will automatically select available matching counters based on the mask.
+        for i in 0..u64::BITS as usize {
+            if counter_mask & (1 << i) != 0 {
+                let _ = self
+                    .hw_counters_info
+                    .get(counter_index + i)
+                    .ok_or(SbiError::InvalidParam)?;
+            }
```

Comment:
If the purpose of this is to check that all bits in `counter_mask` are in range and not necessarily that they correspond to implemented counters, can't you just check the MSB in `counter_mask` rather than looping over all 64 bits?

e.g.

```
if (u64::BITS - counter_mask.leading_zeros()) + counter_index > self.hw_counters_info.len() {
    return Err(SbiError::InvalidParam);
}
```

Reply:
Done.